docs/contributing/999-rfcs/2024-06-02-ts-client.md

```diff
@@ -0,0 +1,298 @@
+const client = new Client();
+
+await client.bucket.put("key", "value");
+```
+
+The client code is the same as the generated javascript as if it were used within wing itself!
+
+## Implementation Details
+
+### Client context data
+
+The client context data is a JSON object that contains all the necessary information to connect to the resources.
+For now, the only information that needs to be stored are environment variables.
+This corresponds to the only API available on `IInflightHost`, which is `addEnvironment`, which is how these values are added:
+
+```json
+{
+  "version": "1",
+  "env": {
+    "BUCKET_HANDLE_a107cec5": "abcxyz"
+  }
+}
+```
+
+The client itself is effectively a proxy of an unseen/unavailable `IInflightHost`, so any expressivity that interface has must be handled by the client.
+
+### Permissions
+
+The client does not directly/automatically handle permissions needed for the arbitrary host it could run on.
+There is no expectation that the deployer of the host is also using wing so there is no shared system to transfer permission knowledge.
+
+For example, when using
+
+```wing
+client.lift(lift { bucket: [put] });
+```
+
+it would be helpful to retrieve the necessary information about the `bucket` and the `put` operation.
+
+If you're using terraform and deploying to AWS, you will need the bucket name and the correspond IAM policy.
+There is no abstract API available to store/retrieve this information currently as this is handled only by target-specific subtypes of `IInflightHost`.
+
+To keep this simple, we can provide and a helper API in the shared aws module (and other platforms, as needed) to generate the necessary policies by exposing the lift information (i.e. the `_liftMap`) from the client.
+
+```wing
+bring aws;
+
+let policies = aws.generatePolicyStatements(client.lifts);
+```
+
+With these, one can generate a policy document to attach to a role/user to be used by the client's host.
+
+If you're using the simulator, this problem is (currently) more difficult.
+The simulator is centralized and will have no knowledge of the future clients that will be using it.
+The simulator has no API to add new permissions after the simulator has started and the scope of that feels too large for this RFC.
+
+As an MVP, the simulator can be configured to ignore permissions via `wing.toml`.
+Even though this is unsafe and cause cause local/cloud differences, wing is intentionally not claiming ownership of the client's host environment.
+
+### Type generation with `lift`
+
+Wing's compiler has type information available for use, but only before preflight execution occurs.
+This makes an API like the following not feasible to implement using logic similar to the existing `extern` or `@inflight`:
+
+```wing
+client.lift("bucket", bucket, ["put"]);
+```
+
+While executing preflight, the knowledge about the type of the `bucket` instance is lost, so generating the actual client code will have no way to reference it.
```

Comment:
> What if we simply had the Wing compiler generates a toInflightDts() method on every generated class that returns a string of its TypeScript DTS types?

This implies we would need to eagerly generate .d.ts for all types, which is a waste of compiler time when not used. This would have to include any referenced JSII types too, since the generated .d.ts files will need to reference them.

Additionally, the final types need to know what has been lifted. If you only lift `put` on bucket, the entire bucket client type should not be available.

Reply:
> This implies we would need to eagerly generate .d.ts for all types, which is a waste of compiler time when not used. This would have to include any referenced JSII types too, since the generated .d.ts files will need to reference them.

Maybe I got confused -- I saw the RFC says "An initial implementation of this can re-use the existing .d.ts generation, and the std.LiftQualificationsContainer can store a path to that generated file to be used as-is when generating the full client in preflight". Does this mean we will still be generating .d.ts files in wingc for every `lift { }` expression? Or would it mean it's being generated during preflight runtime (sounds like it could be slower than doing it in compiled Rust)?

> Additionally, the final types need to know what has been lifted. If you only lift put on bucket, the entire bucket client type should not be available.

Don't we handle this in the TypeScript framework already? with the Pick/Omit stuff?

Reply:
> Does this mean we will still be generating .d.ts files in wingc for every lift { } expression?

Yes. The important part of the `lift` syntax is that our compiler can analyze the types and ops statically and emit the precise type information during justification to be used later. This would not be possible with an api like `client.lift(bucket, ["put"])` without a relatively rigorous reflection system to do the same thing.

A big part of the motivation for direction is that I think the changes to `lift` can be used elsewhere, like unifying how lifting works in `@inflight` and resources like `cloud.Website`

> Don't we handle this in the TypeScript framework already? with the Pick/Omit stuff?

It does, but the context is different. That is operating on fully packed winglibs in preflight.
The clients are inflight-only, and should ideally not even contain any irrelevant types at all. Additionally, the clients will not even have dependencies since they'll ideally be bundled like all our other inflight code (which I should probably mention in this do, oopsie).

My hunch is that getting `lift` to work here would be less work (both implementation and design-wise) than getting reflection to work well enough to support the use case.

Reply:
Ok, got it. yeah it would be cool to see a sketch showing more of how std.LiftQualificationsContainer would look like but it sounds like it could be handy